doc/subsystems/codegen/modules.rst

```diff
@@ -0,0 +1,154 @@
+*********************************
+
+::
+
+    codegen.import_module('dts_basic')
+
+
+This section is deprecated in current pull request
+
+
+
+Device declare module
+*********************
+
+::
+
+    codegen.import_module('declare')
+
+
+The module provides a simple API for driver devices instantiation
+
+Declaring a driver
+------------------
+
+.. function:: declare.device_declare(compatibles, init_prio_flag, kernel_level, irq_func, init_func, api, data_struct, config_struct)
+
+Generate device instances code for all devices activated ('status' = 'ok') in board device tree file
+matching the provided compatibles.
+Most of the parameters provided aim at filling DEVICE_AND_API_INIT macro. Other parameters are there
+to help code generation to fit driver specifics.
+Generated instance code will be set under Kconfig variable which name is build with device node label name
+(e.g: CONFIG_I2C_1).
+
+    :compatibles: List of compatibles supported by the driver (e.g. ['st,stm32-spi-fifo', 'st,stm32-spi'])
+    :init_prio_flag: Flag for driver activation priority (e.g. CONFIG_KERNEL_INIT_PRIORITY_DEVICE)
+    :kernel_level: Flag for driver activation priority (e.g. POST_KERNEL)
+    :irq_func: Two elements python dict providing driver isr function prefix (e.g. 'irq_func':'stm32_i2c_isr')
+                     and flag to be used for driver IRQ code control (e.g. 'irq_flag': 'CONFIG_I2C_STM32_INTERRUPT').
+                     If irq_func is 'None', no IRQ code is generated.
+                     'device_declare' will  generate as much IRQ code as declared by device node.
+                     If 'interrupts-names' property if provided in node, isr names will be generated using matching
+                     values as function postfixs.
+    :init_func: Name of the driver init function (e.g. 'i2c_stm32_init').
+    :api: Name of the driver api structure  (e.g. 'api_funcs').
+    :data_struct: Two elements python list providing elements for driver '_data' structure generation.
+    :config_struct: Two elements python list providing elements for driver '_config' structure generation.
+
+'data_struct' and 'config_struct' will be processed the same way:
+
+* First element (e.g. 'i2c_stm32_config') should be the structure name.
+* Second element is a 'c' template code between triple double quotes (""" """). It should provide the expected code to be generated for the structure. For instance:
+
+.. code-block:: python
+
+    """
+    	.i2c = (I2C_TypeDef *)${@reg/0},
+    	.pclken = {
+    			.enr = ${@clock/bits},
+    			.bus = ${@clock/bus},
+    	},
+    #ifdef CONFIG_I2C_STM32_INTERRUPT
+    	.irq_config_func = st_stm32_i2c_v1_${node_index}_config_irq,
+    #endif
+    	.bitrate = ${@clock-frequency},
+    """
+
+This templates features two types of placeholders that we be substituted by expected matching values:
+
+* ${node_index}: will be replaced by device instance node property 'label' if provided or numeric index otherwise.
```

Comment:
Why don´t call it ${driver-name} and be consistent to what it is named in the currrent Zephyr INIT macro for drivers?

To use an index is the fallback strategy if there is no ´label´ property.